### PR TITLE
Feature: Implement KMeans attribution support and TakesMost LRP rule

### LINKED ISSUE #198: Neuralized K-Means: make k-means amenable to neural network explanations
## Description
K-Means finds some centroids $\mathbf{\mu}_1,\ldots,\mathbf{\mu}_K$ and assigns data points to a cluster as $c = {\rm argmin}_k \lbrace \Vert\boldsymbol{x} - \mathbf{\mu}_k\Vert^2\rbrace$, or in code:

```
c = torch.argmin(torch.cdist(x, centroids)**2)
```

Neither the assignment $c$ nor the distance $\Vert\boldsymbol{x} - \mathbf{\mu}_c\Vert^2$ are really explainable.
The assignment $c$ is not continuous, and the distance is in fact measuring a dissimilarity to the cluster, essentially the opposite of what we want to explain.

## Fixes
[From Clustering to Cluster Explanation via Neural Networks (2022)](https://ieeexplore.ieee.org/stamp/stamp.jsp?arnumber=9817459) present a method based on neuralization. They show that the k-means cluster assignment can be exactly rewritten as a set of piecewise linear functions 
```math
f_c(\boldsymbol{x}) = \min_{k\neq c}\lbrace \boldsymbol{w}_{ck}^\top\boldsymbol{x} + b_{ck}\rbrace
```
and the original k-means cluster assignment can be recovered as $c = {\rm argmax}_k\lbrace f_k(\boldsymbol{x})\rbrace$.

The cluster discriminant $f_c$ is a measure of cluster membership (instead of a dissimilarity) and is also structurally amenable to LRP and friends. It can also be _plugged_ on top of a neural network feature extractor to make deep cluster assignments explainable.

## Additional Information
- There is no _pairwise distance_ layer in PyTorch. Since zennit canonizers apply to `torch.nn.Module` modules, such module would need to be added
- Each discriminant $f_c$ has it's own weight matrix $W\in\mathbb{R}^{(K-1)\times D}$. In principle, a neuralized k-means layer could be written as a layer 
```
torch.nn.Sequential(
    torch.nn.Linear(in_features=D, out_features=K*(K-1)), 
    torch.nn.Unflatten(1, torch.Size([K, K-1]))
)
```
or alternatively via `torch.einsum`, which I think is easier. Either way, a special layer is needed for that.
- The paper suggests a specific LRP rule for the $\min_{k\neq c}\lbrace s_k\rbrace$ to make the explanations continuous. In practice, we can replace it by $\beta^{-1}{\rm LogMeanExp}_k\lbrace \beta\cdot s_k\rbrace$ which approximates the min for $\beta < 0$, but has continuous gradients. Gradient propagation through this layer recovers exactly the LRP rule from the paper, but this also provides explanation continuity for other gradient-based explanation methods. I would suggest to implement such a `LogMeanExpPool` layer. Alternative could be to implement a `MinPool` layer with a custom LRP rule.
- A canonizer could be used to replace the distance/k-means layer by it's neuralized counterpart. Alternatively, a composite might also work, but it seems a bit more complicated (where are the weights computed? how to set LRP rules for the linear/einsum layer?)

### PR DESCRIPTION
**Pull request by jackmcrider**
This commit introduces several new components to enhance k-means based model attribution and provide a new LRP rule for pooling layers:

1. New Layers (`src/zennit/layer.py`):
   - `PairwiseCentroidDistance`: Computes pairwise distances between inputs and a set of centroids.
   - `NeuralizedKMeans`: A layer representing k-means discriminants as a linear transformation (tensor-matrix product with bias).
   - `MinPool1d` and `MinPool2d`: Min-pooling layers for 1D and 2D inputs, implemented by negating inputs/outputs of MaxPool.

2. New Canonizer (`src/zennit/canonizers.py`):
   - `KMeansCanonizer`: Replaces `PairwiseCentroidDistance` layers (with `power=2`) with a sequence of `NeuralizedKMeans`, `MinPool1d`, and `torch.nn.Flatten`. This neuralization enables the application of LRP rules to k-means like clustering outputs.

3. New LRP Rule (`src/zennit/rules.py`):
   - `TakesMost`: An LRP rule designed for max-pooling layers (and by extension, min-pooling layers after neuralization).
   - Distributes relevance based on a softmax of input contributions within each pooling window, providing a "soft" alternative to winner-takes-all.
   - Implements window-wise maximum subtraction for robust numerical stability during the softmax calculation.
   - Dynamically adapts internal convolution, pooling, and transposed convolution operations to match the input dimensionality (1D/2D).
   - Utilizes common, refactored parameters for unpooling operations.
   - Respects the original module's hyperparameters.

4. New Tests (`tests/`):
   - `test_kmeans_canonizer` in `test_canonizers.py`: Verifies the correctness of the `KMeansCanonizer`, ensuring module replacement, functional equivalence of cluster assignments, and proper restoration.

Replaces #197 
Closes #198

### CODE REVIEW

**Comment by jackmcrider** [comment]
Some tests fail, but I think these are mostly lint errors coming from the main branch.

**Comment by chr5tphr** [comment]
Hey Jacob,

Thanks a lot for finalizing the PR! I will get rid of the tests on Python 3.7, so that's fine.
The pylint errors are also fine, I do not think the number of positional arguments is too bad.
The flake8 is valid, you can just add a whitespace where reported.

Could you go through the places that are missing in the coverage report? Some might be unrelated to KMeans and TakesMost, so you can ignore those. It would be good to have those places at least called in some test.

```text
 Name                       Stmts   Miss Branch BrPart  Cover   Missing
----------------------------------------------------------------------
src/zennit/canonizers.py     144      2     34      1    98%   381-382
src/zennit/composites.py     128      0     34      1    99%   159->161
src/zennit/layer.py           27      1      0      0    96%   157
src/zennit/rules.py          138     43     14      0    69%   455-460, 464, 468, 471, 475-482, 486-499, 503-507, 510-564
----------------------------------------------------------------------
TOTAL                       1123     46    312      2    96%
```

**Comment by chr5tphr** [comment]
Just a heads-up @jackmcrider : I am planning to go over this again, fix potential issues, and merge once #215 is merged. Ideally, I will be done by today.

**Comment by jackmcrider** [comment]
I did not see your reply until now, sorry. You mean more tests, right? I will provide more, but not today. Thank you for considering this PR!

**Comment by chr5tphr** [comment]
Thanks! I might also look into it, as I am pushing Zennit this week for a new release. I have some other things still on the list, but would be happy to get this PR into the release.

Some building infrastructure changed, but the building and testing should still work with the same commands. Be sure to rebase.
